### DISCUSSION TITLE
Upload user avatars and store in attributes as URL instead of base64 encoded data

### DISCUSSION THREAD

**Discussion by drpetersen** [goauthentik/authentik#6824] [Show and tell]
# Background

There is an issue (#2631) ongoing about how to allow users to upload their own avatars. Basic functionality seems to work with the file prompt field (#3156) and the `AUTHENTIK_AVATARS=attributes.avatar,…` [setting](https://goauthentik.io/docs/installation/configuration#authentik_avatars), but a few problems still remain to be solved.

While a file cropper seems nice to have, the show-stopper for me is having to store image files as base64-encoded strings directly in the user profile (attributes). Even with reasonably sized image files, this leads to performance issues (e.g., see #5942), makes it more cumbersome to edit the user profile in the admin interface (the YAML is hundreds of lines long), and may cause problems with proxying authentik-ated web applications (with avatars of ca. 100 KBytes, I received error 500 codes from nginx until I found out that I had to increase proxy buffers in size); even then, the `/outpost.goauthentik.io/sign_out` redirect for proxied applications errored out because query strings (presumably containing user profile data) became too long (I suppose).

My proposed solution is a workaround for this problem which allows the user to upload an image file as a user avatar, stores this image file in your authentik server's `/media/` subdirectory under a random (non-guessable) file name, and stores the URL of this image file in the `avatar` profile attribute of the user.

This way, arbitrarily large image files (the maximum is configurable) can be used as a user avatar without cluttering the profile.

Credits: The helpful comments in issue #4283 as well as the above-mentioned issues pointed me in the right direction.

**UPDATE 2026-01-16**: The original code no longer works as of authentik 2025.12. I updated this post to include the fix provided in the discussion below. Thanks to all who contributed, in particular to @krom, @WilliamSoderberg, and to @nicedevil007 for wrapping this up to a [complete working policy script](https://github.com/goauthentik/authentik/discussions/6824#discussioncomment-15511187).

# Procedure

* Include the `attributes.avatar` option in your `AUTHENTIK_AVATARS` environment variable ([see here](https://goauthentik.io/docs/installation/configuration#authentik_avatars)).
* Make sure the `/media/` directory in your authentik-server container is writable by authentik and persistent. Avatar image files will be stored in a sub-directory ~`/media/user-avatars/` (auto-created if not present)~ `/media/public/user-avatars/` which needs to be created with [appropriate permissions](https://github.com/goauthentik/authentik/discussions/6824#discussioncomment-9809766).
* Create a custom `File` prompt, e.g., `default-user-settings-field-avatar`, with field key `attributes.avatar`; adjust the `Order` (not visible in the screenshot) as appropriate (e.g., 205)
![Screenshot_20230910_181841](https://github.com/goauthentik/authentik/assets/9879249/0929f7ff-a3a5-4a90-a959-c6732d6babb0)

* Create a custom `checkbox` prompt, e.g., `default-user-settings-field-avatar-reset`, with field key `attributes.avatar_reset`; adjust the `Order` (e.g., to 206). The user can check this box to delete the current avatar file.
* Create a custom expression policy, e.g., `default-user-settings-avatar-authorization` with ~[this expression](https://gist.github.com/drpetersen/c73b1362264c056c02f3ff8973d55d70)~ the expression provided in @nicedevil007 's [comment below](https://github.com/goauthentik/authentik/discussions/6824#discussioncomment-15511187). This is where the uploaded file is saved, any previous file is removed, and the new URL is prepared to be stored in the user profile. Make sure to configure `AK_DOMAIN`, `MAX_UPLOAD_SIZE` and `ACCEPTED_FILE_TYPES` to your needs.

![Screenshot_20230910_184103](https://github.com/goauthentik/authentik/assets/9879249/9d4db91c-977a-46ad-8dfe-dc9297d9a3e3)

* Update your `default-user-settings` prompt stage (or a new, custom prompt stage, if you prefer) so that the `default-user-settings-field-avatar` and `default-user-settings-field-avatar-reset` prompts are included in the Fields selection (make sure not to accidentally unselect the default profile fields) and the `default-user-settings-avatar-authorization` is included in the Validation Policies (along with the stock `default-user-settings-authorization` policy).
![Screenshot_20230910_183318](https://github.com/goauthentik/authentik/assets/9879249/58ec4970-f812-4e5d-8d65-a639f86afca6)

* Done! Now your users should be able to customize their own avatar images. (Don't forget to re-create your docker container after updating environment options.)

# Open Questions/Issues

* I am not sure if I am simply missing something, and this would have been possible with much less effort (at this time). And maybe there are more efficient ways to achieve this.
* I have not quite thought through the security implications of exposing the user avatar via a world-readable URL, protected only by its random name (a UUID). My intuition is that anyone who can see the avatar would also be allowed to know its URL, as they could copy the image file anyway, even if its download was protected. But maybe the name should be generated in a more cryptographically secure way, and/or the avatar URL should be behind authentik's access control.
* This solution relies on some Python packages being available in expression policies, which might not be the same in future authentik versions.
* The `default-user-settings-avatar-authorization` *should* catch and reject too large files, but there seems to be an implicit limit where things start to go wrong. Specifically, I had problems with large (but still reasonably sized) files: After uploading a files of 1 – 2 MB, I get no error message, but the UI switches from direct display of profile fields to a button "Open settings", which then opens the profile on a separate page, and when I hit the "Continue" button, the upload seems to have failed (silently). I don't know where to begin debugging this. -- EDIT: This is solved, I forgot to increase the `client_max_body_size` setting in my nginx reverse proxy. :roll_eyes: 
* Finally, I was not sure if this is the right place to put this. Issue #2631 seems to be headed in a different direction, issue #4283 had already been closed. So, I put this here in the hope that someone might find it helpful. :smile:

**Comment by thefeli73** [comment]
for me it did not in fact autocreate the `user-avatars/` folder, i had to create it manually for this to work. Great guide otherwise :D

**Comment by Elara6331** [comment]
Thank you so much for this, it works very well! I just had to add the following if statement to the beginning of `remove_old_avatar_file()` to make this policy work during enrollment when there is no existing user yet:

```python
if not request.user.is_authenticated:
  return
```

**Reply by dasunsrule32** [reply]
This works for normal usage, but when impersonation is used, it breaks things pretty badly.

**Reply by masterwishx** [reply]
> Thank you so much for this, it works very well! I just had to add the following if statement to the beginning of `remove_old_avatar_file()` to make this policy work during enrollment when there is no existing user yet:
> 
> ```python
> if not request.user.is_authenticated:
>   return
> ```

is this is right ?

```
def remove_old_avatar_file():
  if not request.user.is_authenticated:    <----- here
  return                                                   <------|
  avatar = request.user.attributes.get("avatar", None)
  if avatar:
    components = avatar.split(URL_PATH, 1)
    if len(components) == 2 and components[0] == "" and components[1]:
      old_filename = FILE_PATH + components[1]
      if isfile(old_filename):
        remove(old_filename)

**Reply by masterwishx** [reply]
> This works for normal usage, but when impersonation is used, it breaks things pretty badly.

can you explain ?

**Reply by dasunsrule32** [reply]
Yeah, enable it and try to impersonate. It's been awhile since I've checked it, but if I recall correctly, you can never leave impersonation with it enabled. You have to actually completely logout and back in.

**Reply by masterwishx** [reply]
Sorry but what this mean impersonate?

**Reply by masterwishx** [reply]
OK i got what you mean , Thanks

**Reply by masterwishx** [reply]
> Yeah, enable it and try to impersonate. It's been awhile since I've checked it, but if I recall correctly, you can never leave impersonation with it enabled. You have to actually completely logout and back in.

yep i got access denied

**Comment by Aetherinox** [comment]
This guide works great. Only thing to note is as @thefeli73 said, the `user-avatars` folder is not automatically created and if you attempt to upload an avatar, it'll throw an error until you create the folder and assign proper perms / owner.

```shell
mkdir -p /authentik/media/user-avatars/
chown -R 1000:1000 /authentik/media/user-avatars/
chmod -R 644 /authentik/media/user-avatars/
```

Obviously adjust chown owner to whatever you need. I got the avatar upload to work with permission `644`

<br />

And if you have multiple sources for avatars, stick the custom upload to the front, otherwise initials will have priority. Mine didn't work until I flipped the order.
```
attributes.avatar,gravatar,initials
```

![image](https://github.com/user-attachments/assets/ab816107-662e-4ce0-99b9-5b4ee956b526)

**Comment by masterwishx** [comment]
cant add validation policies , after success got blank again

**Reply by masterwishx** [reply]
seems its added but not shown in update prompt stage

**Comment by wrenix** [comment]
Could we get an blueprint ?
For a more stable setup. Reproducible and so on.

**Comment by 250king** [comment]
I am thinking is it possible that we can make a expression to upload the image to some storage service such as AWS S3?
But there still some problem need to solve: how to use the non-standard library?

**Reply by rissson** [reply]
```
import boto3
client = boto3.client("s3")
```
and go from there with https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html

**Reply by 250king** [reply]
Is the library included in Authentik defaultly?
In fact, I am using Alibaba Cloud oss, so I need ```oss2``` or ```requests```.

**Reply by rissson** [reply]
requests is included, boto3 as well. You can check https://github.com/goauthentik/authentik/blob/main/poetry.lock for a list of included dependencies and versions.

**Reply by 250king** [reply]
Thanks! In fact Alibaba Cloud OSS support S3 protocol. ```boto3``` is enough. I will try to write a expression.

**Comment by UnInfamous** [comment]
Authentik's [documentation](https://goauthentik.io/docs/installation/configuration#authentik_avatars) in the Background surrounding `AUTHENTIK_AVATARS=attributes.avatar,…`, is no longer there.

I followed these steps and confirmed again that I followed this setup exactly (minus different names for the policy/promts), however, it does show "Avatar image" as seen below:
![image](https://github.com/user-attachments/assets/bf0cad5b-463a-4c4f-b389-74f38e1066f9)


I don't see the image on the profile within Authentik. I can upload the file (and it successfully goes to `/user-avatars/`), also confirmed ordering attributes.avatar to be first of the list of types of avatar images. I've also tried different PNGs & JPGs. Other than adding Apps/Providers, I haven't really changed much of anything within Authentik.

I'm new to identity providers, so any help on what I might have missed would be great.

**Reply by nikolaioak** [reply]
It looks like the avatar settings were removed in this commit: https://github.com/goauthentik/authentik/commit/96554de17a8af90d6a9197c4b65dd177f5caea2b. The instructions for avatar options are [here](https://docs.goauthentik.io/docs/sys-mgmt/settings#avatars) now, but I don't see anything regarding the environment variable previously included in the install config section. I was going to give this method a stab today to see how it works, but the removal of the environment variable has me wondering if this method is no longer supported.

**Reply by nikolaioak** [reply]
Regardless this actually worked for me, set it up in a few minutes. Avatar is displaying as expected - potentially a bit warped based on the aspect ratio of the picture. I would verify you edited the authorization expression policy properly with your environment's details.

**Comment by wrenix** [comment]
Is there a blueprint to add the Prompt and so on for a automatic and reproduce able setup

**Comment by drashna** [comment]
the humanize python package appears to no longer be available in the 2025.8.0-rc5 preview.

**Comment by krom** [comment]
My temporary workaround for missing `humanize` module is there:

1. comment import of module
```python
# from humanize import naturalsize
```

2. add own implementation of `naturalsize`
```python
def naturalsize(value, binary=False, format="%.1f"):
    """
    Convert a size in bytes into a human-readable format.

    :param value: size in bytes
    :param binary: if True, use 1024-based units (KiB, MiB), otherwise 1000-based (KB, MB)
    :param format: number format string (default "%.1f")
    :return: string like "10.5 MB"
    """
    units = ["B", "KB", "MB", "GB", "TB", "PB", "EB"]
    if binary:
        step = 1024.0
        units = ["B", "KiB", "MiB", "GiB", "TiB", "PiB", "EiB"]
    else:
        step = 1000.0

    size = float(value)
    for unit in units:
        if abs(size) < step:
            return (format % size).rstrip("0").rstrip(".") + " " + unit
        size /= step
    return (format % size).rstrip("0").rstrip(".") + " " + units[-1]
```

**Reply by SolninjaA** [reply]
Thank you SO much.

**Comment by KiyoNetcat** [comment]
fyi this now seems to be outright broken on 2025.12
even after adjusting the paths in the user and script to account for the new storage changes, the avatar fails to load with a 404 error
looking at the application icons, apparently there's now a `token` query parameter for it to load the icon, and if that parameter is missing, it returns a 404 for those icons, despite their presence if you included the `token`.

**Reply by nicedevil007** [reply]
just migrated today and all my user avatars are gone :(
already played around with the path and didn't got it working

**Reply by tomorrow-nf** [reply]
Confirming this as well. All avatars are just placeholder text now. Potentially related, my custom CSS that was working previously is also now completely ignored.

**Reply by WilliamSoderberg** [reply]
I managed to get it to work by replicating the internal way they sign the JWT token. 

Here is the code to generate the url or more relevant the token:
```python
AK_DOMAIN = "yourdomain.com"
USAGE = "media"
SCHEMA = "public"
SUB_PATH = "user-avatars" # Your custom subfolder
INTERNAL_WRITE_DIR = f"/{USAGE}/{SCHEMA}/{SUB_PATH}/"
JWT_PATH_PREFIX = f"{USAGE}/{SCHEMA}/{SUB_PATH}/"
URL_BASE = f"https://{AK_DOMAIN}/files/{JWT_PATH_PREFIX}"

def generate_signed_url(filename):
    """Replicates Authentik's internal JWT signing logic exactly."""
    base_secret = environ.get("AUTHENTIK_SECRET_KEY", "")
    
    signing_key = hashlib.sha256(f"{base_secret}:{USAGE}".encode()).hexdigest()
    
    path_claim = f"{JWT_PATH_PREFIX}{filename}"
    
    now = int(time.time())
    payload = {
        "path": path_claim,
        "exp": now + (10 * 365 * 24 * 60 * 60), # 10 Year
        "nbf": now - 30,
    }
    
    token = jwt.encode(
        payload, 
        signing_key, 
        algorithm="HS256"
    )
    
    if isinstance(token, bytes):
        token = token.decode("utf-8")
        
    return f"{URL_BASE}{filename}?token={token}"
```

**Reply by nicedevil007** [reply]
> I managed to get it to work by replicating the internal way they sign the JWT token.
> 
> Here is the code to generate the url or more relevant the token:


Thank you very much for this input, I was able to get it up and running with this and I added features from original idea back (check filetype and size f.e.)

```python
# Authentik User Avatar Policy with JWT-signed URLs
# For Enrollment/User Settings Stage

from uuid import uuid4
from base64 import b64decode
from os import remove, environ
from os.path import isfile
from urllib.parse import urlparse, parse_qs
import hashlib
import time
import jwt

# === CONFIGURATION ===
AK_DOMAIN = "auth.domain.tld"
USAGE = "media"
SCHEMA = "public"
SUB_PATH = "user-avatars"

INTERNAL_WRITE_DIR = f"/data/{USAGE}/{SCHEMA}/{SUB_PATH}/"
JWT_PATH_PREFIX = f"{USAGE}/{SCHEMA}/{SUB_PATH}/"
URL_BASE = f"https://{AK_DOMAIN}/files/{JWT_PATH_PREFIX}"

MAX_UPLOAD_SIZE = 5 * 1024 * 1024  # 5 MB
ACCEPTED_FILE_TYPES = {
    "image/png": "png",
    "image/jpeg": "jpeg",
    "image/webp": "webp",
    "image/svg+xml": "svg",
    "image/gif": "gif",
    "image/bmp": "bmp",
    "image/vnd.microsoft.icon": "ico"
}
EMPTY_FILE = "data:application/octet-stream;base64,"

# Token validity: 10 years
TOKEN_EXPIRY_SECONDS = 10 * 365 * 24 * 60 * 60


# === HELPER FUNCTIONS ===

def format_size(size_bytes):
    """Converts bytes to human-readable size (KiB, MiB, etc.)"""
    for unit in ['B', 'KiB', 'MiB', 'GiB']:
        if size_bytes < 1024.0:
            return f"{size_bytes:.0f} {unit}"
        size_bytes /= 1024.0
    return f"{size_bytes:.0f} TiB"


def generate_signed_url(filename):
    """Generates a JWT-signed URL for file access."""
    base_secret = environ.get("AUTHENTIK_SECRET_KEY", "")
    signing_key = hashlib.sha256(f"{base_secret}:{USAGE}".encode()).hexdigest()
    
    path_claim = f"{JWT_PATH_PREFIX}{filename}"
    now = int(time.time())
    
    payload = {
        "path": path_claim,
        "exp": now + TOKEN_EXPIRY_SECONDS,
        "nbf": now - 30,
    }
    
    token = jwt.encode(payload, signing_key, algorithm="HS256")
    
    if isinstance(token, bytes):
        token = token.decode("utf-8")
    
    return f"{URL_BASE}{filename}?token={token}"


def extract_filename_from_avatar_url(avatar_url):
    """Extracts the filename from a signed avatar URL."""
    if not avatar_url or not avatar_url.startswith(URL_BASE):
        return None
    
    # Parse URL and extract path (without query parameters)
    parsed = urlparse(avatar_url)
    path = parsed.path
    
    # Filename is the last part of the path
    if "/" in path:
        return path.rsplit("/", 1)[-1]
    return None


def remove_old_avatar_file():
    """Deletes the user's old avatar file if it exists."""
    avatar = request.user.attributes.get("avatar", None)
    if not avatar:
        return
    
    filename = extract_filename_from_avatar_url(avatar)
    if filename:
        old_filepath = INTERNAL_WRITE_DIR + filename
        if isfile(old_filepath):
            try:
                remove(old_filepath)
            except OSError:
                pass  # Ignore deletion errors


# === MAIN LOGIC ===

prompt_data = request.context.get("prompt_data")
avatar_overwritten = False

if "avatar" in prompt_data.get("attributes", {}):
    avatar = prompt_data["attributes"]["avatar"]
    
    if avatar == EMPTY_FILE:
        # No file uploaded, ignore field
        del prompt_data["attributes"]["avatar"]
    else:
        # 1. Validate MIME type
        try:
            avatar_mimetype = avatar.split("data:", 1)[1].split(";", 1)[0]
        except (IndexError, AttributeError):
            ak_message("Invalid file format for user avatar.")
            return False
        
        if avatar_mimetype not in ACCEPTED_FILE_TYPES:
            allowed_types = ", ".join(ACCEPTED_FILE_TYPES.values())
            ak_message(f"User avatar must be an image file ({allowed_types}).")
            return False
        
        # 2. Decode Base64 and validate size
        try:
            avatar_base64 = avatar.split(",", 1)[1]
            avatar_binary = b64decode(avatar_base64)
        except (IndexError, ValueError):
            ak_message("Avatar file could not be decoded.")
            return False
        
        avatar_size = len(avatar_binary)
        if avatar_size > MAX_UPLOAD_SIZE:
            ak_message(f"User avatar file size must not exceed {format_size(MAX_UPLOAD_SIZE)}.")
            return False
        
        # 3. Generate random filename
        file_extension = ACCEPTED_FILE_TYPES[avatar_mimetype]
        avatar_filename = f"{uuid4()}.{file_extension}"
        
        # 4. Save file
        full_path = INTERNAL_WRITE_DIR + avatar_filename
        try:
            with open(full_path, "wb") as f:
                f.write(avatar_binary)
        except OSError as e:
            ak_message(f"Failed to save avatar file: {type(e).__name__}: {e}")
            return False
        except Exception as e:
            ak_message(f"Unexpected error: {type(e).__name__}: {e}")
            return False
        
        # 5. Delete old file and set new signed URL
        avatar_overwritten = True
        remove_old_avatar_file()
        prompt_data["attributes"]["avatar"] = generate_signed_url(avatar_filename)


# Handle avatar reset
if "avatar_reset" in prompt_data.get("attributes", {}):
    del prompt_data["attributes"]["avatar_reset"]
    
    # If a new avatar was uploaded, the previous one is already deleted
    if not avatar_overwritten:
        remove_old_avatar_file()
        prompt_data["attributes"]["avatar"] = None
        ak_message("User avatar has been deleted.")

return True
```

**Reply by dpw13** [reply]
Perhaps obvious, but if you store your `AUTHENTIK_SECRET_KEY` as a docker secret by specifying a `file:///` URI for your environment variable make sure to update the above script to read the secret file rather than the environment variable itself.

**Reply by drpetersen** [reply]
Thank you all very much! This is brilliant. I updated the original post to refer to @nicedevil007's working code. :tada:

**Reply by dasunsrule32** [reply]
Should this pick up the existing avatars?

**Reply by dasunsrule32** [reply]
I attempted to update the `avatar` key for some of the users and it's not working.
```
https://auth.domain.tld/files/media/public/user-avatars/<image-file-name>
```

**Reply by dasunsrule32** [reply]
Users will need to upload a new image as the new `avatar` url now appends the token to the URL. I really wish this would become a first class citizen in authentik.

**Reply by ishioni** [reply]
@dasunsrule32 no you have to re-upload them because the signature has to be regenerated in the attribute

**Reply by l-e-e-o** [reply]
I actually found that it sometimes can be useful to have the base64 version available as well. So, I stored the generated URL in`attributes.avatar-url`. This also makes it so the icon is shown in authentik as well. All modifications:

```python
...

# 5. Delete old file and set new signed URL
        avatar_overwritten = True
        remove_old_avatar_file()
        prompt_data["attributes"]["avatar-url"] = generate_signed_url(avatar_filename)
        prompt_data["attributes"]["avatar-base64"] = avatar_base64

# Handle avatar reset
if "avatar_reset" in prompt_data.get("attributes", {}):
    del prompt_data["attributes"]["avatar_reset"]
    
    # If a new avatar was uploaded, the previous one is already deleted
    if not avatar_overwritten:
        remove_old_avatar_file()
        prompt_data["attributes"]["avatar"] = None
        prompt_data["attributes"]["avatar-url"] = None
        prompt_data["attributes"]["avatar-base64"] = None
        ak_message("User avatar has been deleted.")

return True
```

I then created three property mappings:

`Data-URL Avatar Mapping` (Scope: `avatar`)
```python
return {"avatar": request.user.attributes.get("avatar", None)}
```

`Base64 Avatar Mapping` (Scope: `avatar`)
```python
return {"avatar": request.user.attributes.get("avatar-base64", None)}
```

`URL-based Avatar Mapping` (Scope: `avatar`)
```python
return {"avatar": request.user.attributes.get("avatar-url", None)}
```

This way I can just pick the right binding for each application/provider.

**Reply by masterwishx** [reply]
Do you need to change also : User settings - > Avatars?

**Reply by l-e-e-o** [reply]
Yeah I changed it to `attributes.avatar,gravatar,initials` and authentik is able to show the uploaded avatars.

**Reply by masterwishx** [reply]
Thanks, this is what I have already

**Reply by phil-lipp** [reply]
> I actually found that it sometimes can be useful to have the base64 version available as well. So, I stored the generated URL in`attributes.avatar-url`. This also makes it so the icon is shown in authentik as well. All modifications:
> ...
> This way I can just pick the right binding for each application/provider.

Indeed very handy - thank you so much @l-e-e-o!

**Reply by charredchar** [reply]
I've been trying to get the base64 version working with SSSD but I've been running into an issue.

It seems to be coming in as a string (ldapsearch shows this as `jpegPhoto: `) instead of bytes (ldapsearch shows this as `jpegPhoto:: `) so I can't get it to map properly to the `/var/lib/AccountsService/icons/user` properly.
Has anyone successfully used the above code for this?

I have an Active Directory server running with a base64 encoded jpegPhoto value that does come up in ldapsearch as `jpegPhoto:: `. I haven't tested with my services I currently have working with AD through LDAP to see how they handle Authentik's LDAP base64 value.

**Reply by outarde** [reply]
> I actually found that it sometimes can be useful to have the base64 version available as well. So, I stored the generated URL in`attributes.avatar-url`. This also makes it so the icon is shown in authentik as well. All modifications:

As I understand it, your modification just changes the way the avatar is saved to base64 because the "avatar" attribute is no longer written and the "avatar_url" attribute does not exist. Attribute "avatar-base64" also doesn't matter. The file directly uploaded by the user is written to the "avatar" attribute in the html form, since the uploaded image is initially in this format, and is saved in the "avatar" field as is. Checks will interrupt script execution if the size is exceeded but writing the file to the folder doesn't make sense with this method.

Therefore, if someone uses this method remove the code that uploads the avatar file to the server. You're unlikely to be able to open the avatar file via a direct link anyway, even if you pass it in property mapping for other purposes (I didn't succeed).

At the moment, I see it as possible to display avatar files by direct link only with the properly configured reverse proxy (nginx, etc).

**Comment by ZEROX7** [comment]
is it me or does that make authentik use much more ram? rasb pi 4 4gb crashes after 5min if i upload my avatar
removing my avatar fixes that

**Reply by VGHS-lucaruby** [reply]
<img width="2343" height="1312" alt="image" src="https://github.com/user-attachments/assets/833bcc0b-5323-48aa-97d7-17d86366f874" />
Just checked my logs for you, I got around to finally fixing this on the 10th @ ~12:00. Seems that mostly only the cache increased.
Though Authentik is now using the most RAM it has in the last year.
<img width="2341" height="1357" alt="image" src="https://github.com/user-attachments/assets/0b192406-a8b5-4c94-a465-625f1e12cb2c" />
 This coincided with the instance updating to 2026.2.0 So this is more likely the cause for your RAM increase if you are using this version.
Did also give a restart and the RAM increase stuck. Seems to be using ~500MB more now.


Edit: Just looked up the issues page and see this is a bug and fixed in 2026.2.1
https://github.com/goauthentik/authentik/issues/20537

**Reply by ZEROX7** [reply]
im running 2026.2.1 with no avatar uploaded
<img width="1240" height="177" alt="image" src="https://github.com/user-attachments/assets/74d16d84-4ea0-4a94-9bf5-d6ea09b2c191" />
wehn i rember right i saw server and worker at least 500mb higher after i uploaded my avatar and waited 2min
and i dont know if it would get worse wehn other users on my authentik settup will upload avatars

**Reply by VGHS-lucaruby** [reply]
Decided to upgrade and test on my end for you.
Does seem to be something a little strange after uploading an avatar.
<img width="2361" height="1365" alt="image" src="https://github.com/user-attachments/assets/87d20a9b-3726-46dc-acec-aebe18a19a50" />

Don't have enough time to properly test but hopefully this can help you/someone investigate this a bit.
I would guess its something to do with 2026.2.x updates as this was not an issue with the same script on 2025.12.

**Comment by ishioni** [comment]
For those of you who use S3 as your backingstore, here's a version that works with that, on a private bucket

```
# Authentik User Avatar Policy with S3 Storage
# For Enrollment/User Settings Stage

import io
from base64 import b64decode
from os import environ
from urllib.parse import urlparse
from uuid import uuid4

import boto3
from botocore.config import Config
from botocore.exceptions import BotoCoreError, ClientError

# === CONFIGURATION ===
S3_ENDPOINT = environ.get(
    "AUTHENTIK_STORAGE__S3__ENDPOINT", "https://some.s3.provider"
)
S3_BUCKET = environ.get("AUTHENTIK_STORAGE__S3__BUCKET_NAME", "authentik")
S3_ACCESS_KEY = environ.get("AUTHENTIK_STORAGE__S3__ACCESS_KEY", "")
S3_SECRET_KEY = environ.get("AUTHENTIK_STORAGE__S3__SECRET_KEY", "")

if not S3_ACCESS_KEY or not S3_SECRET_KEY:
    ak_message("S3 credentials are not configured.")
    return False
S3_CUSTOM_DOMAIN = environ.get("AUTHENTIK_STORAGE__S3__CUSTOM_DOMAIN", "")
S3_KEY_PREFIX = "user-avatars/"
S3_URL_EXPIRY = int(environ.get("AUTHENTIK_AVATAR__S3__URL_EXPIRY", 3153600000))  # ~100 years

if S3_CUSTOM_DOMAIN:
    URL_BASE = f"https://{S3_CUSTOM_DOMAIN}/{S3_KEY_PREFIX}"
else:
    URL_BASE = f"{S3_ENDPOINT.rstrip('/')}/{S3_BUCKET}/{S3_KEY_PREFIX}"

MAX_UPLOAD_SIZE = 5 * 1024 * 1024  # 5 MB
ACCEPTED_FILE_TYPES = {
    "image/png": "png",
    "image/jpeg": "jpeg",
    "image/webp": "webp",
    "image/svg+xml": "svg",
    "image/gif": "gif",
    "image/bmp": "bmp",
    "image/vnd.microsoft.icon": "ico",
}
EMPTY_FILE = "data:application/octet-stream;base64,"

# === S3 CLIENT ===
s3_client = boto3.client(
    "s3",
    endpoint_url=S3_ENDPOINT,
    aws_access_key_id=S3_ACCESS_KEY,
    aws_secret_access_key=S3_SECRET_KEY,
    config=Config(
        signature_version="s3v4",
        s3={"addressing_style": "virtual" if S3_CUSTOM_DOMAIN else "path"},
    ),
)


# === HELPER FUNCTIONS ===


def format_size(size_bytes):
    """Converts bytes to human-readable size (KiB, MiB, etc.)"""
    for unit in ["B", "KiB", "MiB", "GiB"]:
        if size_bytes < 1024.0:
            return f"{size_bytes:.0f} {unit}"
        size_bytes /= 1024.0
    return f"{size_bytes:.0f} TiB"


def generate_avatar_url(filename):
    """Generates a presigned URL for the avatar on S3."""
    return s3_client.generate_presigned_url(
        "get_object",
        Params={"Bucket": S3_BUCKET, "Key": f"{S3_KEY_PREFIX}{filename}"},
        ExpiresIn=S3_URL_EXPIRY,
    )


def extract_filename_from_avatar_url(avatar_url):
    """Extracts the filename from an S3 avatar URL."""
    if not avatar_url or not avatar_url.startswith(URL_BASE):
        return None

    # Parse URL and extract path (without query parameters)
    parsed = urlparse(avatar_url)
    path = parsed.path

    # Filename is the last part of the path
    if "/" in path:
        return path.rsplit("/", 1)[-1]
    return None


def upload_avatar_to_s3(filename, avatar_binary, content_type):
    """Uploads avatar binary data to S3."""
    s3_key = f"{S3_KEY_PREFIX}{filename}"
    try:
        s3_client.put_object(
            Bucket=S3_BUCKET,
            Key=s3_key,
            Body=io.BytesIO(avatar_binary),
            ContentType=content_type,
        )
    except ClientError as e:
        ak_message(f"Failed to upload avatar: {e.response['Error']['Message']}")
        return False
    except BotoCoreError as e:
        ak_message(f"Failed to upload avatar: {e}")
        return False
    except Exception as e:
        ak_message(f"Unexpected error uploading avatar: {type(e).__name__}: {e}")
        return False
    return True


def delete_old_avatar():
    """Deletes the user's old avatar from S3 if it exists."""
    avatar = request.user.attributes.get("avatar", None)
    if not avatar:
        return

    filename = extract_filename_from_avatar_url(avatar)
    if filename:
        s3_key = f"{S3_KEY_PREFIX}{filename}"
        try:
            s3_client.delete_object(Bucket=S3_BUCKET, Key=s3_key)
        except (ClientError, BotoCoreError):
            pass  # Ignore deletion errors


# === MAIN LOGIC ===

prompt_data = request.context.get("prompt_data")
avatar_overwritten = False

if "avatar" in prompt_data.get("attributes", {}):
    avatar = prompt_data["attributes"]["avatar"]

    if avatar == EMPTY_FILE:
        # No file uploaded, ignore field
        del prompt_data["attributes"]["avatar"]
    else:
        # 1. Validate MIME type
        try:
            avatar_mimetype = avatar.split("data:", 1)[1].split(";", 1)[0]
        except (IndexError, AttributeError):
            ak_message("Invalid file format for user avatar.")
            return False

        if avatar_mimetype not in ACCEPTED_FILE_TYPES:
            allowed_types = ", ".join(ACCEPTED_FILE_TYPES.values())
            ak_message(f"User avatar must be an image file ({allowed_types}).")
            return False

        # 2. Decode Base64 and validate size
        try:
            avatar_base64 = avatar.split(",", 1)[1]
            avatar_binary = b64decode(avatar_base64)
        except (IndexError, ValueError):
            ak_message("Avatar file could not be decoded.")
            return False

        avatar_size = len(avatar_binary)
        if avatar_size > MAX_UPLOAD_SIZE:
            ak_message(
                f"User avatar file size must not exceed {format_size(MAX_UPLOAD_SIZE)}."
            )
            return False

        # 3. Generate random filename
        file_extension = ACCEPTED_FILE_TYPES[avatar_mimetype]
        avatar_filename = f"{uuid4()}.{file_extension}"

        # 4. Upload to S3
        if not upload_avatar_to_s3(avatar_filename, avatar_binary, avatar_mimetype):
            return False

        # 5. Delete old avatar and set new presigned URL
        avatar_overwritten = True
        delete_old_avatar()
        prompt_data["attributes"]["avatar"] = generate_avatar_url(avatar_filename)


# Handle avatar reset
if "avatar_reset" in prompt_data.get("attributes", {}):
    del prompt_data["attributes"]["avatar_reset"]

    # If a new avatar was uploaded, the previous one is already deleted
    if not avatar_overwritten:
        delete_old_avatar()
        prompt_data["attributes"]["avatar"] = None
        ak_message("User avatar has been deleted.")

return True
```

**Comment by nunyab87** [comment]
For anyone who like me, kept having permissions issues no matter what i tried. You can delete the user-avatars directory (if you've already created it) and instead have Authentik auto create it with the right permissions by uploading a file to the files store using Authentik's admin console, make sure to set a custom filename e.g. user-avatars/test.

**Comment by patschi** [comment]
I'm typically against self-promoting, but I genuinly think some might be interested: While I'm limited fan of vibe-coding, it was fine for my personal usage and carefully made "[akvatar](https://github.com/patschi/akvatar)" as an external application (via OIDC) to update the avatar - [screenshots here](https://github.com/patschi/akvatar/blob/master/docs/screenshots.md). I would consider it battle-tested by now.

For example I use `https://authentik.example.com/manage-photo/` for the application, and `https://authentik.example.com/user-avatars/` to proxy any uploaded user-avatars. Both folders are just proxied to akvatar via nginx instead of hitting authentik.

If you don't care, feel free to ignore it. Just wanted to share it anyway, as I find the native avatar integration very lacking.

**Comment by dasunsrule32** [comment]
Just a heads up, this is no longer configurable in the User portal on `2026.8.0`. I haven't dug into it as of now.

**Reply by lastsamurai26** [reply]
If I'm seeing this correctly, the “default-user-settings-avatar-authorization” policy and the “custom-default-user-settings-field-avatar” and “custom-default-user-settings-field-avatar-reset” prompts just need to be re-attached to the "
default-user-settings" stage.


So, under Flows and Stages -> Stages, look for the “default-user-settings” stage, add “custom-default-user-settings-field-avatar” and “custom-default-user-settings-field-avatar-reset” under Fields, and add “default-user-settings-avatar-authorization” under Validation Policies 

(Yes, my prompts and policies have different names)

 In any case, it's working for me again without any problems

**Comment by krom** [comment]
Apparently, this is part of the work on implementing a new mechanism for object attributes, but currently attributes only support 3 data types and do not support customization.
